File: euredact-python/src/euredact/cache.py
"""LRU result cache with SHA-256 content hashing."""

from __future__ import annotations

import hashlib
import threading
from collections import OrderedDict

from euredact.types import RedactResult

# Default ceiling on retained characters, not entries. A count-only cap of
# 1024 entries says nothing about memory: 1024 cached 1 MB documents is a
# gigabyte of PII-bearing text held live. ~16M characters is a few tens of MB.
DEFAULT_MAX_CHARS = 16_000_000


def _result_chars(result: RedactResult) -> int:
    """Approximate retained size of a cached result, in characters."""
    return len(result.redacted_text) + sum(len(d.text) for d in result.detections)


class ResultCache:
    """LRU cache keyed on SHA-256 of input text + config hash.

    Bounded by both entry count and total retained characters — see
    :data:`DEFAULT_MAX_CHARS`.
    """

    def __init__(
        self,
        maxsize: int = 1024,
        enabled: bool = True,
        max_chars: int = DEFAULT_MAX_CHARS,
    ) -> None:
        self._maxsize = maxsize
        self._max_chars = max_chars
        self._enabled = enabled
        self._store: OrderedDict[str, RedactResult] = OrderedDict()
        self._sizes: dict[str, int] = {}
        self._chars = 0
        self._lock = threading.Lock()

    def key(self, text: str, countries: tuple[str, ...], mode: str) -> str:
        """Compute cache key from input text and configuration."""
        # Fed to the hash in pieces rather than joined into one f-string: the
        # f-string copied the whole document before hashing it.
        h = hashlib.sha256()
        h.update(text.encode())
        h.update(b"|")
        h.update("|".join(sorted(countries)).encode())
        h.update(b"|")
        h.update(mode.encode())
        return h.hexdigest()

    def get(self, key: str) -> RedactResult | None:
        """Retrieve a cached result, or None on miss."""
        if not self._enabled:
            return None
        with self._lock:
            if key in self._store:
                self._store.move_to_end(key)
                return self._store[key]
        return None

    def put(self, key: str, result: RedactResult) -> None:
        """Store a result in the cache."""
        if not self._enabled:
            return
        size = _result_chars(result)
        with self._lock:
            self._evict(key)
            # A single result larger than the whole budget would evict
            # everything else and still not fit; skip it rather than empty the
            # cache for one document.
            if size > self._max_chars:
                return
            self._store[key] = result
            self._sizes[key] = size
            self._chars += size
            while self._store and (
                len(self._store) > self._maxsize or self._chars > self._max_chars
            ):
                oldest = next(iter(self._store))
                self._evict(oldest)

    def _evict(self, key: str) -> None:
        """Drop *key* and its accounting. Caller holds the lock."""
        if key in self._store:
            del self._store[key]
        self._chars -= self._sizes.pop(key, 0)

    def clear(self) -> None:
        """Clear the entire cache."""
        with self._lock:
            self._store.clear()
            self._sizes.clear()
            self._chars = 0

File: euredact-ts/src/cache.ts
import { createHash } from "node:crypto";
import type { RedactResult } from "./types.js";

// Default ceiling on retained characters, not entries. A count-only cap of
// 1024 entries says nothing about memory: 1024 cached 1 MB documents is a
// gigabyte of PII-bearing text held live. ~16M characters is a few tens of MB.
const DEFAULT_MAX_CHARS = 16_000_000;

/** Approximate retained size of a cached result, in characters. */
function resultChars(result: RedactResult): number {
  let chars = result.redactedText.length;
  for (const d of result.detections) chars += d.text.length;
  return chars;
}

export class ResultCache {
  private maxsize: number;
  private maxChars: number;
  private enabled: boolean;
  private store = new Map<string, RedactResult>();
  private sizes = new Map<string, number>();
  private chars = 0;

  constructor(maxsize = 1024, enabled = true, maxChars = DEFAULT_MAX_CHARS) {
    this.maxsize = maxsize;
    this.maxChars = maxChars;
    this.enabled = enabled;
  }

  key(text: string, countries: string[], mode: string): string {
    const sorted = [...countries].sort();
    // SHA-256, matching the Python implementation. The previous dual FNV-1a
    // was non-cryptographic and collidable by construction: in a shared-process
    // deployment an attacker who can guess a victim's document could submit a
    // same-length colliding one and be served the victim's cached result,
    // detections and raw matched PII included.
    return createHash("sha256")
      .update(text)
      .update("|")
      .update(sorted.join("|"))
      .update("|")
      .update(mode)
      .digest("hex");
  }

  get(key: string): RedactResult | null {
    if (!this.enabled) return null;
    const result = this.store.get(key);
    if (result !== undefined) {
      // Move to end (LRU)
      this.store.delete(key);
      this.store.set(key, result);
      return result;
    }
    return null;
  }

  put(key: string, result: RedactResult): void {
    if (!this.enabled) return;
    const size = resultChars(result);
    // A single result bigger than the whole budget would evict everything else
    // and still not fit; skip it rather than empty the cache for one document.
    if (size > this.maxChars) {
      this.evict(key);
      return;
    }
    this.evict(key);
    this.store.set(key, result);
    this.sizes.set(key, size);
    this.chars += size;
    while (this.store.size > this.maxsize || this.chars > this.maxChars) {
      const first = this.store.keys().next().value;
      if (first === undefined) break;
      this.evict(first);
    }
  }

  private evict(key: string): void {
    const size = this.sizes.get(key);
    if (size !== undefined) {
      this.chars -= size;
      this.sizes.delete(key);
    }
    this.store.delete(key);
  }

  clear(): void {
    this.store.clear();
    this.sizes.clear();
    this.chars = 0;
  }
}
